{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process MAF file from MC3 to get sample-specific information\n",
    "\n",
    "We want to know if samples with mutations in tumor suppressors have mono-allelic or biallelic knockouts. Our mutation data comes from [the MC3 project](https://gdc.cancer.gov/about-data/publications/mc3-2017), which aggregates mutation calls for TCGA samples from several different calling algorithms and scoring approaches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import gzip\n",
    "from pathlib import Path\n",
    "\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "mc3_filename = Path('./data', 'mc3.v0.2.8.PUBLIC.maf.gz')\n",
    "\n",
    "mutated_samples_dir = Path('./data', 'mutated_samples')\n",
    "mutated_samples_dir.mkdir(exist_ok=True)\n",
    "\n",
    "# gene to get/save mutation info for\n",
    "gene = 'ARID1A'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0               Hugo_Symbol\n",
      "1            Entrez_Gene_Id\n",
      "2                    Center\n",
      "3                NCBI_Build\n",
      "4                Chromosome\n",
      "5            Start_Position\n",
      "6              End_Position\n",
      "7                    Strand\n",
      "8    Variant_Classification\n",
      "9              Variant_Type\n",
      "dtype: object\n"
     ]
    }
   ],
   "source": [
    "mc3 = gzip.open(mc3_filename, \"rb\")\n",
    "maf_header = mc3.readline().decode('UTF-8').strip().split()\n",
    "maf_ixs = {name: ix for ix, name in enumerate(maf_header)}\n",
    "\n",
    "print(pd.Series(maf_header).head(10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "generating mutated samples from MC3 maf file\n"
     ]
    }
   ],
   "source": [
    "mutated_samples_file = mutated_samples_dir / '{}_mutated_samples.tsv'.format(gene)\n",
    "\n",
    "if mutated_samples_file.is_file():\n",
    "    print('file already exists, loading from file')\n",
    "    mutants_df = pd.read_csv(mutated_samples_file, sep='\\t', index_col=0)\n",
    "else:\n",
    "    print('generating mutated samples from MC3 maf file')\n",
    "    mutants = []\n",
    "    for line in mc3:\n",
    "        record = line.decode('UTF-8').strip().split(\"\\t\")\n",
    "        hugo_symbol = record[maf_ixs['Hugo_Symbol']] # gene name\n",
    "        tcga_id_raw = record[maf_ixs['Tumor_Sample_Barcode']] # tumor barcode\n",
    "        tcga_id_raw_normal = record[maf_ixs['Matched_Norm_Sample_Barcode']] # normal barcode\n",
    "        is_tumor = tcga_id_raw.split(\"-\")[3].startswith(\"01\")\n",
    "        tss_code = tcga_id_raw.split(\"-\")[1]\n",
    "\n",
    "        if hugo_symbol == gene:\n",
    "            mutants.append(record)\n",
    "    mutants_df = pd.DataFrame(mutants, columns=maf_header)\n",
    "    mutants_df.to_csv(mutated_samples_file, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1220, 114)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Center</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele1</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>dbSNP_RS</th>\n",
       "      <th>dbSNP_Val_Status</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>Matched_Norm_Sample_Barcode</th>\n",
       "      <th>Match_Norm_Seq_Allele1</th>\n",
       "      <th>Match_Norm_Seq_Allele2</th>\n",
       "      <th>Tumor_Validation_Allele1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>27105667</td>\n",
       "      <td>27105667</td>\n",
       "      <td>+</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>novel</td>\n",
       "      <td>.</td>\n",
       "      <td>TCGA-05-4382-01A-01D-1931-08</td>\n",
       "      <td>TCGA-05-4382-10A-01D-1265-08</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>27089688</td>\n",
       "      <td>27089688</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>TCGA-05-4396-01A-21D-1855-08</td>\n",
       "      <td>TCGA-05-4396-10A-01D-1855-08</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>27056307</td>\n",
       "      <td>27056307</td>\n",
       "      <td>+</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>novel</td>\n",
       "      <td>.</td>\n",
       "      <td>TCGA-05-4425-01A-01D-1753-08</td>\n",
       "      <td>TCGA-05-4425-10A-01D-1753-08</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>27056169</td>\n",
       "      <td>27056169</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>TCGA-05-4427-01A-21D-1855-08</td>\n",
       "      <td>TCGA-05-4427-10A-01D-1855-08</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>27088651</td>\n",
       "      <td>27088651</td>\n",
       "      <td>+</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>TCGA-05-5715-01A-01D-1625-08</td>\n",
       "      <td>TCGA-05-5715-10A-01D-1625-08</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol Entrez_Gene_Id Center NCBI_Build Chromosome Start_Position  \\\n",
       "0      ARID1A              0      .     GRCh37          1       27105667   \n",
       "1      ARID1A              0      .     GRCh37          1       27089688   \n",
       "2      ARID1A              0      .     GRCh37          1       27056307   \n",
       "3      ARID1A              0      .     GRCh37          1       27056169   \n",
       "4      ARID1A              0      .     GRCh37          1       27088651   \n",
       "\n",
       "  End_Position Strand Variant_Classification Variant_Type Reference_Allele  \\\n",
       "0     27105667      +      Nonsense_Mutation          SNP                G   \n",
       "1     27089688      +      Missense_Mutation          SNP                G   \n",
       "2     27056307      +      Nonsense_Mutation          SNP                C   \n",
       "3     27056169      +      Missense_Mutation          SNP                A   \n",
       "4     27088651      +      Nonsense_Mutation          SNP                C   \n",
       "\n",
       "  Tumor_Seq_Allele1 Tumor_Seq_Allele2 dbSNP_RS dbSNP_Val_Status  \\\n",
       "0                 G                 T    novel                .   \n",
       "1                 G                 T        .                .   \n",
       "2                 C                 T    novel                .   \n",
       "3                 A                 C        .                .   \n",
       "4                 C                 T        .                .   \n",
       "\n",
       "           Tumor_Sample_Barcode   Matched_Norm_Sample_Barcode  \\\n",
       "0  TCGA-05-4382-01A-01D-1931-08  TCGA-05-4382-10A-01D-1265-08   \n",
       "1  TCGA-05-4396-01A-21D-1855-08  TCGA-05-4396-10A-01D-1855-08   \n",
       "2  TCGA-05-4425-01A-01D-1753-08  TCGA-05-4425-10A-01D-1753-08   \n",
       "3  TCGA-05-4427-01A-21D-1855-08  TCGA-05-4427-10A-01D-1855-08   \n",
       "4  TCGA-05-5715-01A-01D-1625-08  TCGA-05-5715-10A-01D-1625-08   \n",
       "\n",
       "  Match_Norm_Seq_Allele1 Match_Norm_Seq_Allele2 Tumor_Validation_Allele1  \n",
       "0                      G                      G                        .  \n",
       "1                      G                      G                        .  \n",
       "2                      C                      C                        .  \n",
       "3                      A                      A                        .  \n",
       "4                      C                      C                        .  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(mutants_df.shape)\n",
    "mutants_df.iloc[:5, :20]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:lure_fn]",
   "language": "python",
   "name": "conda-env-lure_fn-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
