{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download multiple modalities of pan-cancer data from TCGA\n",
    "\n",
    "The data is accessed directly from the [Genome Data Commons](https://gdc.cancer.gov/about-data/publications/pancanatlas).\n",
    "\n",
    "NOTE: this download script uses the `md5sum` shell utility to verify file hashes. This script was developed and tested on a Linux machine, and `md5sum` commands may have to be changed to work on other platforms."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from urllib.request import urlretrieve\n",
    "\n",
    "import sys; sys.path.append('..')\n",
    "import config as cfg"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we load a manifest file containing the GDC API ID and filename for each relevant file, as well as the md5 checksum to make sure the whole/uncorrupted file was downloaded.\n",
    "\n",
    "The manifest included in this GitHub repo was downloaded from https://gdc.cancer.gov/node/971 on December 1, 2020."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>filename</th>\n",
       "      <th>md5</th>\n",
       "      <th>size</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>mirna_sample</th>\n",
       "      <td>55d9bf6f-0712-4315-b588-e6f8e295018e</td>\n",
       "      <td>PanCanAtlas_miRNA_sample_information_list.txt</td>\n",
       "      <td>02bb56712be34bcd58c50d90387aebde</td>\n",
       "      <td>553408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>methylation_27k</th>\n",
       "      <td>d82e2c44-89eb-43d9-b6d3-712732bf6a53</td>\n",
       "      <td>jhu-usc.edu_PANCAN_merged_HumanMethylation27_H...</td>\n",
       "      <td>5cec086f0b002d17befef76a3241e73b</td>\n",
       "      <td>5022150019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>methylation_450k</th>\n",
       "      <td>99b0c493-9e94-4d99-af9f-151e46bab989</td>\n",
       "      <td>jhu-usc.edu_PANCAN_HumanMethylation450.betaVal...</td>\n",
       "      <td>a92f50490cf4eca98b0d19e10927de9d</td>\n",
       "      <td>41541692788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rppa</th>\n",
       "      <td>fcbb373e-28d4-4818-92f3-601ede3da5e1</td>\n",
       "      <td>TCGA-RPPA-pancan-clean.txt</td>\n",
       "      <td>e2b914c7ecd369589275d546d9555b05</td>\n",
       "      <td>18901234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rna_seq</th>\n",
       "      <td>3586c0da-64d0-4b74-a449-5ff4d9136611</td>\n",
       "      <td>EBPlusPlusAdjustPANCAN_IlluminaHiSeq_RNASeqV2....</td>\n",
       "      <td>02e72c33071307ff6570621480d3c90b</td>\n",
       "      <td>1882540959</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    id  \\\n",
       "name                                                     \n",
       "mirna_sample      55d9bf6f-0712-4315-b588-e6f8e295018e   \n",
       "methylation_27k   d82e2c44-89eb-43d9-b6d3-712732bf6a53   \n",
       "methylation_450k  99b0c493-9e94-4d99-af9f-151e46bab989   \n",
       "rppa              fcbb373e-28d4-4818-92f3-601ede3da5e1   \n",
       "rna_seq           3586c0da-64d0-4b74-a449-5ff4d9136611   \n",
       "\n",
       "                                                           filename  \\\n",
       "name                                                                  \n",
       "mirna_sample          PanCanAtlas_miRNA_sample_information_list.txt   \n",
       "methylation_27k   jhu-usc.edu_PANCAN_merged_HumanMethylation27_H...   \n",
       "methylation_450k  jhu-usc.edu_PANCAN_HumanMethylation450.betaVal...   \n",
       "rppa                                     TCGA-RPPA-pancan-clean.txt   \n",
       "rna_seq           EBPlusPlusAdjustPANCAN_IlluminaHiSeq_RNASeqV2....   \n",
       "\n",
       "                                               md5         size  \n",
       "name                                                             \n",
       "mirna_sample      02bb56712be34bcd58c50d90387aebde       553408  \n",
       "methylation_27k   5cec086f0b002d17befef76a3241e73b   5022150019  \n",
       "methylation_450k  a92f50490cf4eca98b0d19e10927de9d  41541692788  \n",
       "rppa              e2b914c7ecd369589275d546d9555b05     18901234  \n",
       "rna_seq           02e72c33071307ff6570621480d3c90b   1882540959  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "manifest_df = pd.read_csv(cfg.data_dir / 'manifest.tsv',\n",
    "                          sep='\\t', index_col=0)\n",
    "manifest_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download gene expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloaded data file already exists, skipping download\n"
     ]
    }
   ],
   "source": [
    "rnaseq_id, rnaseq_filename = manifest_df.loc['rna_seq'].id, manifest_df.loc['rna_seq'].filename\n",
    "url = 'http://api.gdc.cancer.gov/data/{}'.format(rnaseq_id)\n",
    "exp_filepath = cfg.data_dir / rnaseq_filename\n",
    "\n",
    "if not exp_filepath.is_file():\n",
    "    urlretrieve(url, exp_filepath)\n",
    "else:\n",
    "    print('Downloaded data file already exists, skipping download')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "02e72c33071307ff6570621480d3c90b  /home/jake/research/mutation-fn/data/EBPlusPlusAdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.tsv\n"
     ]
    }
   ],
   "source": [
    "md5_sum = !md5sum $exp_filepath\n",
    "print(md5_sum[0])\n",
    "assert md5_sum[0].split(' ')[0] == manifest_df.loc['rna_seq'].md5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download mutation data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloaded data file already exists, skipping download\n"
     ]
    }
   ],
   "source": [
    "mutation_id, mutation_filename = manifest_df.loc['mutation'].id, manifest_df.loc['mutation'].filename\n",
    "url = 'http://api.gdc.cancer.gov/data/{}'.format(mutation_id)\n",
    "mutation_filepath = cfg.data_dir / mutation_filename\n",
    "\n",
    "if not mutation_filepath.is_file():\n",
    "    urlretrieve(url, mutation_filepath)\n",
    "else:\n",
    "    print('Downloaded data file already exists, skipping download')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "639ad8f8386e98dacc22e439188aa8fa  /home/jake/research/mutation-fn/data/mc3.v0.2.8.PUBLIC.maf.gz\n"
     ]
    }
   ],
   "source": [
    "md5_sum = !md5sum $mutation_filepath\n",
    "print(md5_sum[0])\n",
    "assert md5_sum[0].split(' ')[0] == manifest_df.loc['mutation'].md5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download gene set from Park et al. paper\n",
    "\n",
    "We want to download the set of genes analyzed in [Park et al. 2021](https://www.nature.com/articles/s41467-021-27242-3). We are particularly interested in the \"Class 2/3/4\" genes from Figure 1, which are inferred to be \"two-hit\" genes where non-synonymous mutations and CNVs tend to co-occur more often than would be expected by chance."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(453, 17)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Pair</th>\n",
       "      <th>Effect</th>\n",
       "      <th>Pval</th>\n",
       "      <th>FDR</th>\n",
       "      <th>LogFDR</th>\n",
       "      <th>Target</th>\n",
       "      <th>Unnamed: 8</th>\n",
       "      <th>Gene.1</th>\n",
       "      <th>Tissue.1</th>\n",
       "      <th>Pair.1</th>\n",
       "      <th>Effect.1</th>\n",
       "      <th>Pval.1</th>\n",
       "      <th>FDR.1</th>\n",
       "      <th>LogFDR.1</th>\n",
       "      <th>Target.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACVR1</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>ACVR1_UCEC</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.997445</td>\n",
       "      <td>0.681521</td>\n",
       "      <td>0.166514</td>\n",
       "      <td>B_Target</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACVR1</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>ACVR1_UCEC</td>\n",
       "      <td>-0.482699</td>\n",
       "      <td>0.370095</td>\n",
       "      <td>0.571956</td>\n",
       "      <td>0.242630</td>\n",
       "      <td>B_Target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACVR2A</td>\n",
       "      <td>COADREAD</td>\n",
       "      <td>ACVR2A_COADREAD</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.997631</td>\n",
       "      <td>0.681315</td>\n",
       "      <td>0.166646</td>\n",
       "      <td>B_Target</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACVR2A</td>\n",
       "      <td>COADREAD</td>\n",
       "      <td>ACVR2A_COADREAD</td>\n",
       "      <td>0.187531</td>\n",
       "      <td>0.753981</td>\n",
       "      <td>0.780954</td>\n",
       "      <td>0.107369</td>\n",
       "      <td>B_Target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACVR2A</td>\n",
       "      <td>LIHC</td>\n",
       "      <td>ACVR2A_LIHC</td>\n",
       "      <td>-1.000178</td>\n",
       "      <td>0.040008</td>\n",
       "      <td>0.122307</td>\n",
       "      <td>0.912515</td>\n",
       "      <td>B_Target</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACVR2A</td>\n",
       "      <td>LIHC</td>\n",
       "      <td>ACVR2A_LIHC</td>\n",
       "      <td>0.163421</td>\n",
       "      <td>0.834858</td>\n",
       "      <td>0.821548</td>\n",
       "      <td>0.085362</td>\n",
       "      <td>B_Target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AJUBA</td>\n",
       "      <td>HNSC</td>\n",
       "      <td>AJUBA_HNSC</td>\n",
       "      <td>-0.936050</td>\n",
       "      <td>0.008733</td>\n",
       "      <td>0.045816</td>\n",
       "      <td>1.338886</td>\n",
       "      <td>A_Hit</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AJUBA</td>\n",
       "      <td>HNSC</td>\n",
       "      <td>AJUBA_HNSC</td>\n",
       "      <td>0.711518</td>\n",
       "      <td>0.026392</td>\n",
       "      <td>0.147049</td>\n",
       "      <td>0.832509</td>\n",
       "      <td>B_Target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>AKT1_BRCA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.997163</td>\n",
       "      <td>0.681829</td>\n",
       "      <td>0.166318</td>\n",
       "      <td>B_Target</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AKT1</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>AKT1_BRCA</td>\n",
       "      <td>0.535765</td>\n",
       "      <td>0.092176</td>\n",
       "      <td>0.312738</td>\n",
       "      <td>0.504805</td>\n",
       "      <td>B_Target</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene    Tissue             Pair    Effect      Pval       FDR    LogFDR  \\\n",
       "0   ACVR1      UCEC       ACVR1_UCEC  0.000000  0.997445  0.681521  0.166514   \n",
       "1  ACVR2A  COADREAD  ACVR2A_COADREAD  0.000000  0.997631  0.681315  0.166646   \n",
       "2  ACVR2A      LIHC      ACVR2A_LIHC -1.000178  0.040008  0.122307  0.912515   \n",
       "3   AJUBA      HNSC       AJUBA_HNSC -0.936050  0.008733  0.045816  1.338886   \n",
       "4    AKT1      BRCA        AKT1_BRCA  0.000000  0.997163  0.681829  0.166318   \n",
       "\n",
       "     Target  Unnamed: 8  Gene.1  Tissue.1           Pair.1  Effect.1  \\\n",
       "0  B_Target         NaN   ACVR1      UCEC       ACVR1_UCEC -0.482699   \n",
       "1  B_Target         NaN  ACVR2A  COADREAD  ACVR2A_COADREAD  0.187531   \n",
       "2  B_Target         NaN  ACVR2A      LIHC      ACVR2A_LIHC  0.163421   \n",
       "3     A_Hit         NaN   AJUBA      HNSC       AJUBA_HNSC  0.711518   \n",
       "4  B_Target         NaN    AKT1      BRCA        AKT1_BRCA  0.535765   \n",
       "\n",
       "     Pval.1     FDR.1  LogFDR.1  Target.1  \n",
       "0  0.370095  0.571956  0.242630  B_Target  \n",
       "1  0.753981  0.780954  0.107369  B_Target  \n",
       "2  0.834858  0.821548  0.085362  B_Target  \n",
       "3  0.026392  0.147049  0.832509  B_Target  \n",
       "4  0.092176  0.312738  0.504805  B_Target  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "park_df = pd.read_excel(\n",
    "    'https://static-content.springer.com/esm/art%3A10.1038%2Fs41467-021-27242-3/MediaObjects/41467_2021_27242_MOESM11_ESM.xlsx',\n",
    "    sheet_name='SupplFig3', header=1\n",
    ")\n",
    "print(park_df.shape)\n",
    "park_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(79, 7)\n",
      "['AJUBA' 'ARID1A' 'ARID2' 'BAP1' 'BCOR' 'BRD7' 'CDH1' 'CDKN2A' 'CTCF'\n",
      " 'CTNNB1' 'CUL3' 'CYLD' 'EP300' 'EPAS1' 'FAT1' 'FGFR3' 'HRAS' 'KEAP1'\n",
      " 'MAP2K4' 'MAP3K1' 'NF1' 'NF2' 'NRAS' 'NSD1' 'PBRM1' 'PIK3R1' 'PPP2R1A'\n",
      " 'PSIP1' 'PTEN' 'PTPDC1' 'RB1' 'RUNX1' 'SIN3A' 'SMAD4' 'SMARCA4' 'STK11'\n",
      " 'TP53' 'TSC1' 'TSC2' 'VHL' 'ZFHX3' 'ZNF750']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Effect</th>\n",
       "      <th>Pval</th>\n",
       "      <th>FDR</th>\n",
       "      <th>LogFDR</th>\n",
       "      <th>Target</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pair</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AJUBA_HNSC</th>\n",
       "      <td>AJUBA</td>\n",
       "      <td>HNSC</td>\n",
       "      <td>-0.936050</td>\n",
       "      <td>8.732955e-03</td>\n",
       "      <td>0.045816</td>\n",
       "      <td>1.338886</td>\n",
       "      <td>A_Hit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARID1A_LGG</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>LGG</td>\n",
       "      <td>-0.904202</td>\n",
       "      <td>1.956617e-02</td>\n",
       "      <td>0.076291</td>\n",
       "      <td>1.117471</td>\n",
       "      <td>A_Hit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARID1A_STAD</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>STAD</td>\n",
       "      <td>1.133548</td>\n",
       "      <td>3.979932e-04</td>\n",
       "      <td>0.000294</td>\n",
       "      <td>3.517309</td>\n",
       "      <td>B_Target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARID1A_UCEC</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>1.597876</td>\n",
       "      <td>3.451937e-07</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>B_Target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARID2_LIHC</th>\n",
       "      <td>ARID2</td>\n",
       "      <td>LIHC</td>\n",
       "      <td>-0.978419</td>\n",
       "      <td>1.019284e-02</td>\n",
       "      <td>0.050276</td>\n",
       "      <td>1.298552</td>\n",
       "      <td>A_Hit</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Gene Tissue    Effect          Pval       FDR    LogFDR  \\\n",
       "Pair                                                                     \n",
       "AJUBA_HNSC    AJUBA   HNSC -0.936050  8.732955e-03  0.045816  1.338886   \n",
       "ARID1A_LGG   ARID1A    LGG -0.904202  1.956617e-02  0.076291  1.117471   \n",
       "ARID1A_STAD  ARID1A   STAD  1.133548  3.979932e-04  0.000294  3.517309   \n",
       "ARID1A_UCEC  ARID1A   UCEC  1.597876  3.451937e-07  0.000000  5.000000   \n",
       "ARID2_LIHC    ARID2   LIHC -0.978419  1.019284e-02  0.050276  1.298552   \n",
       "\n",
       "               Target  \n",
       "Pair                   \n",
       "AJUBA_HNSC      A_Hit  \n",
       "ARID1A_LGG      A_Hit  \n",
       "ARID1A_STAD  B_Target  \n",
       "ARID1A_UCEC  B_Target  \n",
       "ARID2_LIHC      A_Hit  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# the paper uses an FDR threshold of 0.1, so we do the same here\n",
    "fdr_threshold = 0.1\n",
    "\n",
    "park_loss_df = (park_df\n",
    "  .loc[park_df.FDR < fdr_threshold, :]\n",
    "  .iloc[:, :8]\n",
    "  .set_index('Pair')\n",
    ")\n",
    "print(park_loss_df.shape)\n",
    "print(park_loss_df.Gene.unique())\n",
    "park_loss_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(30, 7)\n",
      "['ARID1A' 'ATRX' 'BRAF' 'CTNNB1' 'CUL3' 'EGFR' 'EPAS1' 'ERBB4' 'FBXW7'\n",
      " 'GATA3' 'IDH1' 'KIT' 'KMT2B' 'KRAS' 'NFE2L2' 'NRAS' 'PDGFRA' 'PPP2R1A'\n",
      " 'TP53']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Effect</th>\n",
       "      <th>Pval</th>\n",
       "      <th>FDR</th>\n",
       "      <th>LogFDR</th>\n",
       "      <th>Target</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pair</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ARID1A_UCEC</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>-2.351526</td>\n",
       "      <td>1.154365e-03</td>\n",
       "      <td>-0.000447</td>\n",
       "      <td>3.340299</td>\n",
       "      <td>B_Target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATRX_LGG</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>LGG</td>\n",
       "      <td>1.440987</td>\n",
       "      <td>4.025141e-08</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>A_Hit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRAF_SKCM</th>\n",
       "      <td>BRAF</td>\n",
       "      <td>SKCM</td>\n",
       "      <td>1.239939</td>\n",
       "      <td>2.042839e-11</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>A_Hit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRAF_THCA</th>\n",
       "      <td>BRAF</td>\n",
       "      <td>THCA</td>\n",
       "      <td>-2.449684</td>\n",
       "      <td>7.509064e-04</td>\n",
       "      <td>-0.000209</td>\n",
       "      <td>3.659041</td>\n",
       "      <td>B_Target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTNNB1_UCEC</th>\n",
       "      <td>CTNNB1</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>-1.257861</td>\n",
       "      <td>1.597968e-02</td>\n",
       "      <td>0.096292</td>\n",
       "      <td>1.016366</td>\n",
       "      <td>B_Target</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Gene Tissue    Effect          Pval       FDR    LogFDR  \\\n",
       "Pair                                                                     \n",
       "ARID1A_UCEC  ARID1A   UCEC -2.351526  1.154365e-03 -0.000447  3.340299   \n",
       "ATRX_LGG       ATRX    LGG  1.440987  4.025141e-08  0.000000  5.000000   \n",
       "BRAF_SKCM      BRAF   SKCM  1.239939  2.042839e-11  0.000000  5.000000   \n",
       "BRAF_THCA      BRAF   THCA -2.449684  7.509064e-04 -0.000209  3.659041   \n",
       "CTNNB1_UCEC  CTNNB1   UCEC -1.257861  1.597968e-02  0.096292  1.016366   \n",
       "\n",
       "               Target  \n",
       "Pair                   \n",
       "ARID1A_UCEC  B_Target  \n",
       "ATRX_LGG        A_Hit  \n",
       "BRAF_SKCM       A_Hit  \n",
       "BRAF_THCA    B_Target  \n",
       "CTNNB1_UCEC  B_Target  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "park_gain_df = (park_df\n",
    "  .loc[park_df['FDR.1'] < fdr_threshold, :]\n",
    "  .iloc[:, 9:]\n",
    ")\n",
    "park_gain_df.columns = park_gain_df.columns.str.replace('.1', '')\n",
    "park_gain_df.set_index('Pair', inplace=True)\n",
    "print(park_gain_df.shape)\n",
    "print(park_gain_df.Gene.unique())\n",
    "park_gain_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download oncogene/tumor suppressor information for Park et al. genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Cancer</th>\n",
       "      <th>classification</th>\n",
       "      <th>Decision</th>\n",
       "      <th>Tissue Frequency</th>\n",
       "      <th>Pancan Frequency</th>\n",
       "      <th>Consensus Score</th>\n",
       "      <th>Correlation adusted score</th>\n",
       "      <th>Novel</th>\n",
       "      <th>Rescue Notes</th>\n",
       "      <th>Note about previous publication</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEY</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ABL1_PANCAN</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>PANCAN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rescued</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.011675</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>Evidence from OncoImpact/DriverNET overlap (SN...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR1_UCEC</th>\n",
       "      <td>ACVR1</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>official</td>\n",
       "      <td>0.053030</td>\n",
       "      <td>0.007490</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR1B_PANCAN</th>\n",
       "      <td>ACVR1B</td>\n",
       "      <td>PANCAN</td>\n",
       "      <td>possible tsg</td>\n",
       "      <td>official</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.010904</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Found in 24132290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR2A_COADREAD</th>\n",
       "      <td>ACVR2A</td>\n",
       "      <td>COADREAD</td>\n",
       "      <td>tsg</td>\n",
       "      <td>official</td>\n",
       "      <td>0.028481</td>\n",
       "      <td>0.013988</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Found in 22810696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR2A_LIHC</th>\n",
       "      <td>ACVR2A</td>\n",
       "      <td>LIHC</td>\n",
       "      <td>possible tsg</td>\n",
       "      <td>official</td>\n",
       "      <td>0.031073</td>\n",
       "      <td>0.013988</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Found in private communication about integrati...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Gene    Cancer classification  Decision  Tissue Frequency  \\\n",
       "KEY                                                                            \n",
       "ABL1_PANCAN        ABL1    PANCAN            NaN   rescued               NaN   \n",
       "ACVR1_UCEC        ACVR1      UCEC       oncogene  official          0.053030   \n",
       "ACVR1B_PANCAN    ACVR1B    PANCAN   possible tsg  official               NaN   \n",
       "ACVR2A_COADREAD  ACVR2A  COADREAD            tsg  official          0.028481   \n",
       "ACVR2A_LIHC      ACVR2A      LIHC   possible tsg  official          0.031073   \n",
       "\n",
       "                 Pancan Frequency  Consensus Score  Correlation adusted score  \\\n",
       "KEY                                                                             \n",
       "ABL1_PANCAN              0.011675              0.0                        NaN   \n",
       "ACVR1_UCEC               0.007490              1.5                        1.5   \n",
       "ACVR1B_PANCAN            0.010904              1.0                        0.0   \n",
       "ACVR2A_COADREAD          0.013988              1.5                        1.5   \n",
       "ACVR2A_LIHC              0.013988              1.5                        1.5   \n",
       "\n",
       "                 Novel                                       Rescue Notes  \\\n",
       "KEY                                                                         \n",
       "ABL1_PANCAN          0  Evidence from OncoImpact/DriverNET overlap (SN...   \n",
       "ACVR1_UCEC           0                                                NaN   \n",
       "ACVR1B_PANCAN        0                                                NaN   \n",
       "ACVR2A_COADREAD      0                                                NaN   \n",
       "ACVR2A_LIHC          0                                                NaN   \n",
       "\n",
       "                                   Note about previous publication  \n",
       "KEY                                                                 \n",
       "ABL1_PANCAN                                                    NaN  \n",
       "ACVR1_UCEC                                                       0  \n",
       "ACVR1B_PANCAN                                    Found in 24132290  \n",
       "ACVR2A_COADREAD                                  Found in 22810696  \n",
       "ACVR2A_LIHC      Found in private communication about integrati...  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# downloaded from https://www.sciencedirect.com/science/article/pii/S009286741830237X\n",
    "# oncogene/TSG predictions for genes/cancer types using 20/20+ classifier\n",
    "class_df = pd.read_excel(\n",
    "    cfg.data_dir / '1-s2.0-S009286741830237X-mmc1.xlsx', \n",
    "    sheet_name='Table S1', index_col='KEY', header=2\n",
    ")\n",
    "class_df.rename(columns={'Tumor suppressor or oncogene prediction (by 20/20+)':\n",
    "                         'classification'},\n",
    "                inplace=True)\n",
    "\n",
    "class_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TSG' 'Oncogene']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Effect</th>\n",
       "      <th>Pval</th>\n",
       "      <th>FDR</th>\n",
       "      <th>LogFDR</th>\n",
       "      <th>Target</th>\n",
       "      <th>classification</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AJUBA_HNSC</th>\n",
       "      <td>AJUBA</td>\n",
       "      <td>HNSC</td>\n",
       "      <td>-0.936050</td>\n",
       "      <td>8.732955e-03</td>\n",
       "      <td>0.045816</td>\n",
       "      <td>1.338886</td>\n",
       "      <td>A_Hit</td>\n",
       "      <td>TSG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARID1A_LGG</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>LGG</td>\n",
       "      <td>-0.904202</td>\n",
       "      <td>1.956617e-02</td>\n",
       "      <td>0.076291</td>\n",
       "      <td>1.117471</td>\n",
       "      <td>A_Hit</td>\n",
       "      <td>TSG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARID1A_STAD</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>STAD</td>\n",
       "      <td>1.133548</td>\n",
       "      <td>3.979932e-04</td>\n",
       "      <td>0.000294</td>\n",
       "      <td>3.517309</td>\n",
       "      <td>B_Target</td>\n",
       "      <td>TSG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARID1A_UCEC</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>1.597876</td>\n",
       "      <td>3.451937e-07</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>B_Target</td>\n",
       "      <td>TSG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARID2_LIHC</th>\n",
       "      <td>ARID2</td>\n",
       "      <td>LIHC</td>\n",
       "      <td>-0.978419</td>\n",
       "      <td>1.019284e-02</td>\n",
       "      <td>0.050276</td>\n",
       "      <td>1.298552</td>\n",
       "      <td>A_Hit</td>\n",
       "      <td>TSG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Gene Tissue    Effect          Pval       FDR    LogFDR  \\\n",
       "AJUBA_HNSC    AJUBA   HNSC -0.936050  8.732955e-03  0.045816  1.338886   \n",
       "ARID1A_LGG   ARID1A    LGG -0.904202  1.956617e-02  0.076291  1.117471   \n",
       "ARID1A_STAD  ARID1A   STAD  1.133548  3.979932e-04  0.000294  3.517309   \n",
       "ARID1A_UCEC  ARID1A   UCEC  1.597876  3.451937e-07  0.000000  5.000000   \n",
       "ARID2_LIHC    ARID2   LIHC -0.978419  1.019284e-02  0.050276  1.298552   \n",
       "\n",
       "               Target classification  \n",
       "AJUBA_HNSC      A_Hit            TSG  \n",
       "ARID1A_LGG      A_Hit            TSG  \n",
       "ARID1A_STAD  B_Target            TSG  \n",
       "ARID1A_UCEC  B_Target            TSG  \n",
       "ARID2_LIHC      A_Hit            TSG  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loss_class_df = (park_loss_df\n",
    "    .merge(class_df.loc[:, ['classification']], left_index=True, right_index=True)\n",
    ")\n",
    "\n",
    "# format oncogene/TSG classification to work with vogelstein genes\n",
    "loss_class_df['classification'] = (\n",
    "    loss_class_df.classification.str.replace('possible ', '')\n",
    "                                    .replace('tsg', 'TSG')\n",
    "                                    .replace('oncogene', 'Oncogene')\n",
    ")\n",
    "\n",
    "print(loss_class_df.classification.unique())\n",
    "loss_class_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TSG' 'Oncogene']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Effect</th>\n",
       "      <th>Pval</th>\n",
       "      <th>FDR</th>\n",
       "      <th>LogFDR</th>\n",
       "      <th>Target</th>\n",
       "      <th>classification</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ARID1A_UCEC</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>-2.351526</td>\n",
       "      <td>1.154365e-03</td>\n",
       "      <td>-0.000447</td>\n",
       "      <td>3.340299</td>\n",
       "      <td>B_Target</td>\n",
       "      <td>TSG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATRX_LGG</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>LGG</td>\n",
       "      <td>1.440987</td>\n",
       "      <td>4.025141e-08</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>A_Hit</td>\n",
       "      <td>TSG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRAF_SKCM</th>\n",
       "      <td>BRAF</td>\n",
       "      <td>SKCM</td>\n",
       "      <td>1.239939</td>\n",
       "      <td>2.042839e-11</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>A_Hit</td>\n",
       "      <td>Oncogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRAF_THCA</th>\n",
       "      <td>BRAF</td>\n",
       "      <td>THCA</td>\n",
       "      <td>-2.449684</td>\n",
       "      <td>7.509064e-04</td>\n",
       "      <td>-0.000209</td>\n",
       "      <td>3.659041</td>\n",
       "      <td>B_Target</td>\n",
       "      <td>Oncogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTNNB1_UCEC</th>\n",
       "      <td>CTNNB1</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>-1.257861</td>\n",
       "      <td>1.597968e-02</td>\n",
       "      <td>0.096292</td>\n",
       "      <td>1.016366</td>\n",
       "      <td>B_Target</td>\n",
       "      <td>Oncogene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Gene Tissue    Effect          Pval       FDR    LogFDR  \\\n",
       "ARID1A_UCEC  ARID1A   UCEC -2.351526  1.154365e-03 -0.000447  3.340299   \n",
       "ATRX_LGG       ATRX    LGG  1.440987  4.025141e-08  0.000000  5.000000   \n",
       "BRAF_SKCM      BRAF   SKCM  1.239939  2.042839e-11  0.000000  5.000000   \n",
       "BRAF_THCA      BRAF   THCA -2.449684  7.509064e-04 -0.000209  3.659041   \n",
       "CTNNB1_UCEC  CTNNB1   UCEC -1.257861  1.597968e-02  0.096292  1.016366   \n",
       "\n",
       "               Target classification  \n",
       "ARID1A_UCEC  B_Target            TSG  \n",
       "ATRX_LGG        A_Hit            TSG  \n",
       "BRAF_SKCM       A_Hit       Oncogene  \n",
       "BRAF_THCA    B_Target       Oncogene  \n",
       "CTNNB1_UCEC  B_Target       Oncogene  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gain_class_df = (park_gain_df\n",
    "    .merge(class_df.loc[:, ['classification']], left_index=True, right_index=True)\n",
    ")\n",
    "\n",
    "# format oncogene/TSG classification to work with vogelstein genes\n",
    "gain_class_df['classification'] = (\n",
    "    gain_class_df.classification.str.replace('possible ', '')\n",
    "                                    .replace('tsg', 'TSG')\n",
    "                                    .replace('oncogene', 'Oncogene')\n",
    ")\n",
    "\n",
    "print(gain_class_df.classification.unique())\n",
    "gain_class_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "loss_class_df.to_csv(cfg.data_dir / 'park_loss_df.tsv', sep='\\t')\n",
    "gain_class_df.to_csv(cfg.data_dir / 'park_gain_df.tsv', sep='\\t')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
