{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import pickle as pkl\n",
    "import itertools as it\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import sys; sys.path.append('..')\n",
    "import config as cfg\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "cfg.de_input_dir.mkdir(parents=True, exist_ok=True)\n",
    "cfg.de_output_dir.mkdir(parents=True, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(10530, 60498)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000000003.14</th>\n",
       "      <th>ENSG00000000005.5</th>\n",
       "      <th>ENSG00000000419.12</th>\n",
       "      <th>ENSG00000000457.13</th>\n",
       "      <th>ENSG00000000460.16</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0047-01</th>\n",
       "      <td>2132</td>\n",
       "      <td>1</td>\n",
       "      <td>552</td>\n",
       "      <td>387</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0055-01</th>\n",
       "      <td>2590</td>\n",
       "      <td>742</td>\n",
       "      <td>2110</td>\n",
       "      <td>493</td>\n",
       "      <td>452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-2483-01</th>\n",
       "      <td>4438</td>\n",
       "      <td>1</td>\n",
       "      <td>2230</td>\n",
       "      <td>735</td>\n",
       "      <td>653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-2485-01</th>\n",
       "      <td>4615</td>\n",
       "      <td>4</td>\n",
       "      <td>1022</td>\n",
       "      <td>629</td>\n",
       "      <td>672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-04-1331-01</th>\n",
       "      <td>6159</td>\n",
       "      <td>10</td>\n",
       "      <td>2752</td>\n",
       "      <td>1173</td>\n",
       "      <td>1091</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 ENSG00000000003.14  ENSG00000000005.5  ENSG00000000419.12  \\\n",
       "sample_id                                                                    \n",
       "TCGA-02-0047-01                2132                  1                 552   \n",
       "TCGA-02-0055-01                2590                742                2110   \n",
       "TCGA-02-2483-01                4438                  1                2230   \n",
       "TCGA-02-2485-01                4615                  4                1022   \n",
       "TCGA-04-1331-01                6159                 10                2752   \n",
       "\n",
       "                 ENSG00000000457.13  ENSG00000000460.16  \n",
       "sample_id                                                \n",
       "TCGA-02-0047-01                 387                 269  \n",
       "TCGA-02-0055-01                 493                 452  \n",
       "TCGA-02-2483-01                 735                 653  \n",
       "TCGA-02-2485-01                 629                 672  \n",
       "TCGA-04-1331-01                1173                1091  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load gene expression counts data\n",
    "all_counts_df = pd.read_csv(cfg.processed_counts_file, sep='\\t', index_col=0)\n",
    "print(all_counts_df.shape)\n",
    "all_counts_df.iloc[:5, :5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(308148, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>positive_prob</th>\n",
       "      <th>identifier</th>\n",
       "      <th>mutation_status</th>\n",
       "      <th>copy_status</th>\n",
       "      <th>status</th>\n",
       "      <th>class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-4E-A92E-01</td>\n",
       "      <td>1.558183e-08</td>\n",
       "      <td>ACVR1_UCEC</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>none</td>\n",
       "      <td>class 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-A5-A0GB-01</td>\n",
       "      <td>2.734236e-237</td>\n",
       "      <td>ACVR1_UCEC</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>one</td>\n",
       "      <td>class 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-A5-A0GJ-01</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>ACVR1_UCEC</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>none</td>\n",
       "      <td>class 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-A5-A0GM-01</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>ACVR1_UCEC</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>none</td>\n",
       "      <td>class 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-A5-A0GW-01</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>ACVR1_UCEC</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>none</td>\n",
       "      <td>class 1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         sample_id  positive_prob  identifier  mutation_status  copy_status  \\\n",
       "0  TCGA-4E-A92E-01   1.558183e-08  ACVR1_UCEC                0          0.0   \n",
       "1  TCGA-A5-A0GB-01  2.734236e-237  ACVR1_UCEC                1          0.0   \n",
       "2  TCGA-A5-A0GJ-01   0.000000e+00  ACVR1_UCEC                0          0.0   \n",
       "3  TCGA-A5-A0GM-01   0.000000e+00  ACVR1_UCEC                0          0.0   \n",
       "4  TCGA-A5-A0GW-01   0.000000e+00  ACVR1_UCEC                0          0.0   \n",
       "\n",
       "  status    class  \n",
       "0   none  class 1  \n",
       "1    one  class 1  \n",
       "2   none  class 1  \n",
       "3   none  class 1  \n",
       "4   none  class 1  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load park et al. initial analysis results here\n",
    "park_info_df = pd.read_csv(\n",
    "    cfg.park_info_file, sep='\\t'\n",
    ")\n",
    "\n",
    "print(park_info_df.shape)\n",
    "park_info_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for now, we'll just run these with a few hand-picked examples\n",
    "# in the future we should run for all examples + do some kind of synthesis\n",
    "class_1_id = 'NF2_KIRP'\n",
    "class_2_id = 'PTEN_UCEC'\n",
    "class_3_id = 'KRAS_COADREAD'\n",
    "class_4_id = 'TP53_BRCA'\n",
    "\n",
    "class_ids = [class_1_id, class_2_id, class_3_id, class_4_id]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NF2_KIRP none one\n",
      "NF2_KIRP none both\n",
      "NF2_KIRP one both\n",
      "PTEN_UCEC none one\n",
      "PTEN_UCEC none both\n",
      "PTEN_UCEC one both\n",
      "KRAS_COADREAD none one\n",
      "KRAS_COADREAD none both\n",
      "KRAS_COADREAD one both\n",
      "TP53_BRCA none one\n",
      "TP53_BRCA none both\n",
      "TP53_BRCA one both\n"
     ]
    }
   ],
   "source": [
    "exp_output = str(cfg.de_input_dir / '{}_{}_{}_counts.tsv')\n",
    "info_output = str(cfg.de_input_dir / '{}_{}_{}_info.tsv')\n",
    "\n",
    "for class_id in class_ids:\n",
    "    gene, cancer_type = class_id.split('_')\n",
    "    for status_1, status_2 in it.combinations(\n",
    "        ['none', 'one', 'both'], 2\n",
    "    ):\n",
    "        \n",
    "        print(class_id, status_1, status_2)\n",
    "        \n",
    "        info_df = park_info_df[\n",
    "            ((park_info_df.identifier == class_id) &\n",
    "             (park_info_df.status.isin([status_1, status_2])))\n",
    "        ].set_index('sample_id')\n",
    "        \n",
    "        samples = (\n",
    "            info_df.index.intersection(all_counts_df.index)\n",
    "        )\n",
    "        exp_df = all_counts_df.loc[samples, :]\n",
    "        info_df = info_df.loc[samples, :]\n",
    "        \n",
    "        exp_df.to_csv(\n",
    "            exp_output.format(class_id, status_1, status_2),\n",
    "            sep='\\t'\n",
    "        )\n",
    "        info_df.status.to_csv(\n",
    "            info_output.format(class_id, status_1, status_2),\n",
    "            sep='\\t'\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "base_dir = str(cfg.de_base_dir)\n",
    "input_dir = str(cfg.de_input_dir)\n",
    "output_dir = str(cfg.de_output_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Loading required package: S4Vectors\n",
      "\n",
      "R[write to console]: Loading required package: stats4\n",
      "\n",
      "R[write to console]: Loading required package: BiocGenerics\n",
      "\n",
      "R[write to console]: Loading required package: parallel\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: IRanges\n",
      "\n",
      "R[write to console]: Loading required package: GenomicRanges\n",
      "\n",
      "R[write to console]: Loading required package: GenomeInfoDb\n",
      "\n",
      "R[write to console]: Loading required package: SummarizedExperiment\n",
      "\n",
      "R[write to console]: Loading required package: MatrixGenerics\n",
      "\n",
      "R[write to console]: Loading required package: matrixStats\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: Biobase\n",
      "\n",
      "R[write to console]: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/home/jake/research/mutation-fn/4_de_analysis/de_input/NF2_KIRP_none_one_counts.tsv\"\n",
      "[1] \"/home/jake/research/mutation-fn/4_de_analysis/de_input/NF2_KIRP_none_one_info.tsv\"\n",
      "[1] \"/home/jake/research/mutation-fn/4_de_analysis/de_input/NF2_KIRP_none_both_counts.tsv\"\n",
      "[1] \"/home/jake/research/mutation-fn/4_de_analysis/de_input/NF2_KIRP_none_both_info.tsv\"\n",
      "[1] \"/home/jake/research/mutation-fn/4_de_analysis/de_input/NF2_KIRP_one_both_counts.tsv\"\n",
      "[1] \"/home/jake/research/mutation-fn/4_de_analysis/de_input/NF2_KIRP_one_both_info.tsv\"\n",
      "[1] \"/home/jake/research/mutation-fn/4_de_analysis/de_input/PTEN_UCEC_none_one_counts.tsv\"\n",
      "[1] \"/home/jake/research/mutation-fn/4_de_analysis/de_input/PTEN_UCEC_none_one_info.tsv\"\n",
      "[1] \"/home/jake/research/mutation-fn/4_de_analysis/de_input/PTEN_UCEC_none_both_counts.tsv\"\n",
      "[1] \"/home/jake/research/mutation-fn/4_de_analysis/de_input/PTEN_UCEC_none_both_info.tsv\"\n",
      "[1] \"/home/jake/research/mutation-fn/4_de_analysis/de_input/PTEN_UCEC_one_both_counts.tsv\"\n",
      "[1] \"/home/jake/research/mutation-fn/4_de_analysis/de_input/PTEN_UCEC_one_both_info.tsv\"\n",
      "[1] \"/home/jake/research/mutation-fn/4_de_analysis/de_input/KRAS_COADREAD_none_one_counts.tsv\"\n",
      "[1] \"/home/jake/research/mutation-fn/4_de_analysis/de_input/KRAS_COADREAD_none_one_info.tsv\"\n",
      "[1] \"/home/jake/research/mutation-fn/4_de_analysis/de_input/KRAS_COADREAD_none_both_counts.tsv\"\n",
      "[1] \"/home/jake/research/mutation-fn/4_de_analysis/de_input/KRAS_COADREAD_none_both_info.tsv\"\n",
      "[1] \"/home/jake/research/mutation-fn/4_de_analysis/de_input/KRAS_COADREAD_one_both_counts.tsv\"\n",
      "[1] \"/home/jake/research/mutation-fn/4_de_analysis/de_input/KRAS_COADREAD_one_both_info.tsv\"\n",
      "[1] \"/home/jake/research/mutation-fn/4_de_analysis/de_input/TP53_BRCA_none_one_counts.tsv\"\n",
      "[1] \"/home/jake/research/mutation-fn/4_de_analysis/de_input/TP53_BRCA_none_one_info.tsv\"\n",
      "[1] \"/home/jake/research/mutation-fn/4_de_analysis/de_input/TP53_BRCA_none_both_counts.tsv\"\n",
      "[1] \"/home/jake/research/mutation-fn/4_de_analysis/de_input/TP53_BRCA_none_both_info.tsv\"\n",
      "[1] \"/home/jake/research/mutation-fn/4_de_analysis/de_input/TP53_BRCA_one_both_counts.tsv\"\n",
      "[1] \"/home/jake/research/mutation-fn/4_de_analysis/de_input/TP53_BRCA_one_both_info.tsv\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i base_dir -i input_dir -i output_dir \n",
    "\n",
    "source(paste0(base_dir, '/de_analysis.R'))\n",
    "\n",
    "# this stuff probably shouldn't be hardcoded, but fine for now\n",
    "identifiers <- c(\n",
    "    'NF2_KIRP',\n",
    "    'PTEN_UCEC',\n",
    "    'KRAS_COADREAD',\n",
    "    'TP53_BRCA'\n",
    ")\n",
    "\n",
    "status_combs <- list()\n",
    "status_combs[[1]] <- c('none', 'one')\n",
    "status_combs[[2]] <- c('none', 'both')\n",
    "status_combs[[3]] <- c('one', 'both')\n",
    "\n",
    "for (i in 1:length(identifiers)) {\n",
    "    identifier <- identifiers[i]\n",
    "    for (j in 1:length(status_combs)) {\n",
    "        status_1 <- status_combs[[j]][1]\n",
    "        status_2 <- status_combs[[j]][2]\n",
    "        counts_file <- paste(input_dir,\n",
    "                             '/',\n",
    "                             identifier,\n",
    "                             '_',\n",
    "                             status_1,\n",
    "                             '_',\n",
    "                             status_2,\n",
    "                             '_counts.tsv',\n",
    "                             sep='')\n",
    "        info_file <- paste(input_dir,\n",
    "                           '/',\n",
    "                           identifier,\n",
    "                           '_',\n",
    "                           status_1,\n",
    "                           '_',\n",
    "                           status_2,\n",
    "                           '_info.tsv',\n",
    "                           sep='')\n",
    "        print(counts_file)\n",
    "        print(info_file)\n",
    "        # get_DE_stats_DESeq(info_file,\n",
    "        #                    counts_file,\n",
    "        #                    paste(identifier, '_', status_1, '_', status_2,\n",
    "        #                          sep=''),\n",
    "        #                    output_dir)\n",
    "        \n",
    "    }\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:mutation_fn]",
   "language": "python",
   "name": "conda-env-mutation_fn-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
