# functions to perform RNA-seq differential expression analysis using DESeq2
# mostly copied from and inspired by:
# https://github.com/greenelab/generic-expression-patterns/blob/master/generic_expression_patterns_modules/DE_analysis.R

library("DESeq2")

get_DE_stats_DESeq <- function(metadata_file,
                               expression_file,
                               experiment_id,
                               output_dir) {

  # This function performs DE analysis using DESeq.
  # Expression data in expression_file are grouped based on metadata_file
  #
  # Arguments
  # ---------
  # metadata_file: str
  #   File containing mapping between sample id and group
  #
  # expression_file: str
  #   File containing gene expression data
  #
  # experiment_id: str
  #   Experiment id used to label saved output filee
  #
  # output_dir: str
  #   Directory to save output files to

  expression_data <- t(as.matrix(read.csv(expression_file, sep="\t", header=TRUE, row.names=1)))
  metadata <- as.matrix(read.csv(metadata_file, sep="\t", header=TRUE, row.names=1))

  print("Checking sample ordering...")
  print(all.equal(colnames(expression_data), rownames(metadata)))

  group <- interaction(metadata[,1])
  mm <- model.matrix(~0 + group)

  # Note about DESeq object
  # Different function calls to create DESeq object depending on the input data.
  # These objects are then passed into [DESeq](https://www.rdocumentation.org/packages/DESeq2/versions/1.12.3/topics/DESeq).
  #
  # * `DESeqDataSetFromTximport()`: For data generated by RSEM, Kallisto, Salmon, Sailfish
  # * `DESeqDataSetFromMatrix()`: For raw count matrix, expects integer values
  # * `DESeqDataSet()`: For RangedSummarizedExperiment (RSE) data, like from recount2
  #
  # Documentation recommends using `tximport` and then calling `DESeqDataSetFromTximport` for RSEM (or other methods listed above)
  # estimated count data in order to get counts corrected for by gene length: https://support.bioconductor.org/p/94003/#94028.
  #
  # [tximport](https://rdrr.io/bioc/tximport/src/R/tximport.R):
  # * imports transcript-level estimates from various external software and optionally summarizes abundances,
  # counts, and transcript lengths to the gene-level (default) or transcript-level matrices.
  #
  # _Input_ a list of files containing:
  # * transcript-level abundances (TPM/FPKM): Normalized read count. Number of reads mapped to a given
  # transcript normalized by transcript length and/or library size.
  # * estimated/expected counts: An estimate of the number of reads derived from a given transcript/isoform
  # or gene. We can only get an estimate for this quantity since reads do not often map uniquely to a single transcript/isoform.
  # * effective lengths: The number of possible start positions for a read within a transcript
  #
  # _Output_ a list of matrices containing:
  # If the user requests just the transcript-level data, then combine data across samples and return.
  # Otherwise summarizes to gene-level using `tx2gene` parameter: a two-column dataframe linking transcript id (column 1) to gene id (column 2).
  # * abundance matrix
  # * count matrix
  # * length matrix
  #
  # For the analysis in notebook: explore_uncorrelated_genes/1a_DE_polyA_vs_ribo.ipynb
  # We are using data generated by RSEM, which returned gene-level summaries.
  # In this case, `tximport` will just combine the gene-level summaries across files into a matrix.
  #
  # The dataset we have is a matrix of estimated counts that were combined across samples. We don't have the individual raw data files output from RSEM,
  # which is needed to call tximport. However, it looks like `tximport` would have been redundant given what we mentioned above.
  # Here we are rounding the combined RSEM estimated counts and using `DESeqDataSetFromMatrix()`.
  # Given the first step of `DESeqDataSetFromTximport` is rounding the counts, we anticipate this step to be ok.
  # https://rdrr.io/bioc/DESeq2/src/R/AllClasses.R.
  #
  # In general, the recommendation to use `tximport` might be necessary for the case where we have transcript-level data.
  #
  # For the analysis in notebook: human_general_analysis/2_identify_generic_genes_pathways.ipynb
  # We are using RangedSummarizedExperiments from recount2.
  # We combined these RSEs in download_recount2_data.R so we cannot use `DESeqDataSet()` which expects SE inputs.
  # Again, we round our combined RSE data and use `DESeqDataSetFromMatrix()`.
  # Looks like rounding is also performed for `DESeqDataSet()` as well.
  print(head(metadata))
  print(group[1:5])
  ddset <- DESeqDataSetFromMatrix(expression_data, colData=metadata, design = ~group)
  deseq_object <- DESeq(ddset, quiet=TRUE)

  # Note parameter settings:
  # `independentFilter=False`: We have turned off the automatic filtering, which
  # filter filter out those tests from the procedure that have no, or little
  # chance of showing significant evidence, without even looking at their test statistic.
  # Typically, this results in increased detection power at the same experiment-wide
  # type I error, as measured in terms of the false discovery rate.
  # cooksCutoff=True (default): Cook's distance as a diagnostic to tell if a single sample
  # has a count which has a disproportionate impact on the log fold change and p-values.
  # These genes are flagged with an NA in the pvalue and padj columns
  deseq_results <- results(deseq_object, independentFiltering=FALSE)
  deseq_results_df <-  as.data.frame(deseq_results)

  # Save summary statistics of DEGs
  out_file = paste(output_dir, "/DE_stats_", experiment_id, ".txt", sep="")
  write.table(deseq_results_df, file = out_file, row.names = T, sep = "\t", quote = F)
}
