{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculate distance between means/medoids of mutation groupings\n",
    "\n",
    "Our goal is to find an unsupervised way of calculating distance/similarity between our mutation groupings (\"none\"/\"one\"/\"both\") which isn't affected by sample size, to the degree that differentially expressed gene count was (see `4_de_analysis` notebooks).\n",
    "\n",
    "Here, we'll try the extremely simple method of:\n",
    "\n",
    "1) taking the n-dimensional mean (centroid) or median (medoid) of each group  \n",
    "2) calculating distance between the centroids and using this to define \"expression similarity\"\n",
    "\n",
    "We'll try this for a few different feature selection/embedding methods, and for both gene expression and RPPA (protein expression) data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import pickle as pkl\n",
    "import itertools as it\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import sys; sys.path.append('..')\n",
    "import config as cfg\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# whether to use expression or rppa data\n",
    "# data_type = 'expression'\n",
    "data_type = 'rppa'\n",
    "\n",
    "# how to calculate centroids, 'mean' or 'median'\n",
    "centroid_method = 'mean'\n",
    "\n",
    "# number of features to subset to, by mean absolute deviation\n",
    "# TODO try this in PCA/UMAP space too\n",
    "subset_mad_feats = 100"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load expression data\n",
    "\n",
    "We'll also subset to the top features by mean absolute deviation, if that option ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(7790, 189)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X1433EPSILON</th>\n",
       "      <th>X4EBP1</th>\n",
       "      <th>X4EBP1_pS65</th>\n",
       "      <th>X4EBP1_pT37T46</th>\n",
       "      <th>X53BP1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5J2-01</th>\n",
       "      <td>-0.494</td>\n",
       "      <td>0.592</td>\n",
       "      <td>0.10500</td>\n",
       "      <td>-0.13500</td>\n",
       "      <td>1.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-PA-A5YG-01</th>\n",
       "      <td>-0.535</td>\n",
       "      <td>0.183</td>\n",
       "      <td>-0.10600</td>\n",
       "      <td>-0.04010</td>\n",
       "      <td>1.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5JV-01</th>\n",
       "      <td>-0.109</td>\n",
       "      <td>0.418</td>\n",
       "      <td>0.02470</td>\n",
       "      <td>-0.72100</td>\n",
       "      <td>1.66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5JT-01</th>\n",
       "      <td>-0.413</td>\n",
       "      <td>0.259</td>\n",
       "      <td>0.00597</td>\n",
       "      <td>0.56300</td>\n",
       "      <td>1.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5JR-01</th>\n",
       "      <td>-0.288</td>\n",
       "      <td>-0.112</td>\n",
       "      <td>0.01940</td>\n",
       "      <td>-0.00267</td>\n",
       "      <td>2.23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 X1433EPSILON  X4EBP1  X4EBP1_pS65  X4EBP1_pT37T46  X53BP1\n",
       "sample_id                                                                 \n",
       "TCGA-OR-A5J2-01        -0.494   0.592      0.10500        -0.13500    1.89\n",
       "TCGA-PA-A5YG-01        -0.535   0.183     -0.10600        -0.04010    1.03\n",
       "TCGA-OR-A5JV-01        -0.109   0.418      0.02470        -0.72100    1.66\n",
       "TCGA-OR-A5JT-01        -0.413   0.259      0.00597         0.56300    1.46\n",
       "TCGA-OR-A5JR-01        -0.288  -0.112      0.01940        -0.00267    2.23"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression_data_file = (\n",
    "    '/home/jake/research/mpmp/data/tcga_expression_matrix_processed.tsv.gz'\n",
    ")\n",
    "expression_sample_info = (\n",
    "    '/home/jake/research/mpmp/data/sample_info/tcga_expression_sample_identifiers.tsv'\n",
    ")\n",
    "\n",
    "rppa_data_file = (\n",
    "    '/home/jake/research/mpmp/data/tcga_rppa_matrix_processed.tsv'\n",
    ")\n",
    "rppa_sample_info = (\n",
    "    '/home/jake/research/mpmp/data/sample_info/tcga_rppa_sample_identifiers.tsv'\n",
    ")\n",
    "\n",
    "if data_type == 'expression':\n",
    "    data_df = pd.read_csv(expression_data_file, sep='\\t', index_col=0)\n",
    "    sample_info_df = pd.read_csv(expression_data_file, sep='\\t', index_col=0)\n",
    "elif data_type == 'rppa':\n",
    "    data_df = pd.read_csv(rppa_data_file, sep='\\t', index_col=0)\n",
    "    sample_info_df = pd.read_csv(rppa_data_file, sep='\\t', index_col=0)\n",
    "    \n",
    "print(data_df.shape)\n",
    "data_df.iloc[:5, :5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['MYH11' 'VHL' 'ECADHERIN' 'ERALPHA' 'CLAUDIN7']\n",
      "(7790, 100)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MYH11</th>\n",
       "      <th>VHL</th>\n",
       "      <th>ECADHERIN</th>\n",
       "      <th>ERALPHA</th>\n",
       "      <th>CLAUDIN7</th>\n",
       "      <th>ACETYLATUBULINLYS40</th>\n",
       "      <th>RICTOR</th>\n",
       "      <th>FASN</th>\n",
       "      <th>CAVEOLIN1</th>\n",
       "      <th>GAPDH</th>\n",
       "      <th>...</th>\n",
       "      <th>MSH2</th>\n",
       "      <th>AKT</th>\n",
       "      <th>LCK</th>\n",
       "      <th>KU80</th>\n",
       "      <th>P53</th>\n",
       "      <th>SMAD1</th>\n",
       "      <th>NCADHERIN</th>\n",
       "      <th>G6PD</th>\n",
       "      <th>BAP1C4</th>\n",
       "      <th>TSC1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5J2-01</th>\n",
       "      <td>1.950</td>\n",
       "      <td>-0.315</td>\n",
       "      <td>-1.400</td>\n",
       "      <td>-1.020</td>\n",
       "      <td>0.141</td>\n",
       "      <td>2.48</td>\n",
       "      <td>0.881</td>\n",
       "      <td>0.2280</td>\n",
       "      <td>2.160</td>\n",
       "      <td>1.550</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.263</td>\n",
       "      <td>0.696</td>\n",
       "      <td>-0.0920</td>\n",
       "      <td>1.68</td>\n",
       "      <td>-0.743</td>\n",
       "      <td>0.0846</td>\n",
       "      <td>-0.24500</td>\n",
       "      <td>-0.5090</td>\n",
       "      <td>-0.427</td>\n",
       "      <td>2.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-PA-A5YG-01</th>\n",
       "      <td>0.602</td>\n",
       "      <td>0.697</td>\n",
       "      <td>-0.223</td>\n",
       "      <td>-0.445</td>\n",
       "      <td>0.517</td>\n",
       "      <td>1.36</td>\n",
       "      <td>0.305</td>\n",
       "      <td>-0.6990</td>\n",
       "      <td>0.762</td>\n",
       "      <td>0.587</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.266</td>\n",
       "      <td>0.543</td>\n",
       "      <td>-0.0364</td>\n",
       "      <td>1.44</td>\n",
       "      <td>-0.133</td>\n",
       "      <td>-0.1770</td>\n",
       "      <td>-0.18100</td>\n",
       "      <td>-0.2940</td>\n",
       "      <td>-0.531</td>\n",
       "      <td>1.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5JV-01</th>\n",
       "      <td>-0.679</td>\n",
       "      <td>-0.301</td>\n",
       "      <td>-1.350</td>\n",
       "      <td>-1.160</td>\n",
       "      <td>0.336</td>\n",
       "      <td>2.22</td>\n",
       "      <td>0.571</td>\n",
       "      <td>-0.2000</td>\n",
       "      <td>0.501</td>\n",
       "      <td>1.960</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.402</td>\n",
       "      <td>0.437</td>\n",
       "      <td>0.2060</td>\n",
       "      <td>1.32</td>\n",
       "      <td>-0.437</td>\n",
       "      <td>0.1150</td>\n",
       "      <td>-0.25100</td>\n",
       "      <td>0.0225</td>\n",
       "      <td>-0.301</td>\n",
       "      <td>2.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5JT-01</th>\n",
       "      <td>-0.140</td>\n",
       "      <td>-0.481</td>\n",
       "      <td>-1.670</td>\n",
       "      <td>-1.210</td>\n",
       "      <td>0.204</td>\n",
       "      <td>2.75</td>\n",
       "      <td>0.141</td>\n",
       "      <td>-0.9400</td>\n",
       "      <td>1.450</td>\n",
       "      <td>0.503</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.611</td>\n",
       "      <td>0.533</td>\n",
       "      <td>0.3390</td>\n",
       "      <td>1.10</td>\n",
       "      <td>-0.749</td>\n",
       "      <td>0.0357</td>\n",
       "      <td>0.04950</td>\n",
       "      <td>0.7790</td>\n",
       "      <td>-0.912</td>\n",
       "      <td>1.51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5JR-01</th>\n",
       "      <td>-1.040</td>\n",
       "      <td>-0.337</td>\n",
       "      <td>-1.230</td>\n",
       "      <td>-1.090</td>\n",
       "      <td>0.265</td>\n",
       "      <td>3.04</td>\n",
       "      <td>0.331</td>\n",
       "      <td>-0.0643</td>\n",
       "      <td>0.708</td>\n",
       "      <td>1.670</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.470</td>\n",
       "      <td>1.190</td>\n",
       "      <td>0.0180</td>\n",
       "      <td>1.15</td>\n",
       "      <td>-0.669</td>\n",
       "      <td>-0.1130</td>\n",
       "      <td>-0.00368</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>-0.631</td>\n",
       "      <td>1.31</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 100 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 MYH11    VHL  ECADHERIN  ERALPHA  CLAUDIN7  \\\n",
       "sample_id                                                     \n",
       "TCGA-OR-A5J2-01  1.950 -0.315     -1.400   -1.020     0.141   \n",
       "TCGA-PA-A5YG-01  0.602  0.697     -0.223   -0.445     0.517   \n",
       "TCGA-OR-A5JV-01 -0.679 -0.301     -1.350   -1.160     0.336   \n",
       "TCGA-OR-A5JT-01 -0.140 -0.481     -1.670   -1.210     0.204   \n",
       "TCGA-OR-A5JR-01 -1.040 -0.337     -1.230   -1.090     0.265   \n",
       "\n",
       "                 ACETYLATUBULINLYS40  RICTOR    FASN  CAVEOLIN1  GAPDH  ...  \\\n",
       "sample_id                                                               ...   \n",
       "TCGA-OR-A5J2-01                 2.48   0.881  0.2280      2.160  1.550  ...   \n",
       "TCGA-PA-A5YG-01                 1.36   0.305 -0.6990      0.762  0.587  ...   \n",
       "TCGA-OR-A5JV-01                 2.22   0.571 -0.2000      0.501  1.960  ...   \n",
       "TCGA-OR-A5JT-01                 2.75   0.141 -0.9400      1.450  0.503  ...   \n",
       "TCGA-OR-A5JR-01                 3.04   0.331 -0.0643      0.708  1.670  ...   \n",
       "\n",
       "                  MSH2    AKT     LCK  KU80    P53   SMAD1  NCADHERIN    G6PD  \\\n",
       "sample_id                                                                       \n",
       "TCGA-OR-A5J2-01 -0.263  0.696 -0.0920  1.68 -0.743  0.0846   -0.24500 -0.5090   \n",
       "TCGA-PA-A5YG-01 -0.266  0.543 -0.0364  1.44 -0.133 -0.1770   -0.18100 -0.2940   \n",
       "TCGA-OR-A5JV-01 -0.402  0.437  0.2060  1.32 -0.437  0.1150   -0.25100  0.0225   \n",
       "TCGA-OR-A5JT-01 -0.611  0.533  0.3390  1.10 -0.749  0.0357    0.04950  0.7790   \n",
       "TCGA-OR-A5JR-01 -0.470  1.190  0.0180  1.15 -0.669 -0.1130   -0.00368  1.0000   \n",
       "\n",
       "                 BAP1C4  TSC1  \n",
       "sample_id                      \n",
       "TCGA-OR-A5J2-01  -0.427  2.25  \n",
       "TCGA-PA-A5YG-01  -0.531  1.11  \n",
       "TCGA-OR-A5JV-01  -0.301  2.01  \n",
       "TCGA-OR-A5JT-01  -0.912  1.51  \n",
       "TCGA-OR-A5JR-01  -0.631  1.31  \n",
       "\n",
       "[5 rows x 100 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if subset_mad_feats is not None:\n",
    "    mad_ranking = (\n",
    "        data_df.mad(axis=0)\n",
    "               .sort_values(ascending=False)\n",
    "    )\n",
    "    top_feats = mad_ranking[:subset_mad_feats].index.astype(str).values\n",
    "    print(top_feats[:5])\n",
    "    data_df = data_df.reindex(top_feats, axis='columns')\n",
    "    \n",
    "print(data_df.shape)\n",
    "data_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load Park et al. \"hit\" data\n",
    "\n",
    "This was collated/formatted in `0_process_park.ipynb`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>class_name</th>\n",
       "      <th>mutation_status</th>\n",
       "      <th>cnv_status</th>\n",
       "      <th>num_hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-3C-AAAU-01</th>\n",
       "      <td>class 4</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>one</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-3C-AALI-01</th>\n",
       "      <td>class 4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-3C-AALJ-01</th>\n",
       "      <td>class 4</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>one</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-3C-AALK-01</th>\n",
       "      <td>class 4</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>one</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-4H-AAAK-01</th>\n",
       "      <td>class 4</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>one</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                class_name  mutation_status  cnv_status num_hits\n",
       "TCGA-3C-AAAU-01    class 4                0           1      one\n",
       "TCGA-3C-AALI-01    class 4                1           1     both\n",
       "TCGA-3C-AALJ-01    class 4                0           1      one\n",
       "TCGA-3C-AALK-01    class 4                0           1      one\n",
       "TCGA-4H-AAAK-01    class 4                0           1      one"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(cfg.distance_gain_info, 'rb') as f:\n",
    "    park_gain_info = pkl.load(f)\n",
    "    \n",
    "park_gain_info['TP53_BRCA'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>class_name</th>\n",
       "      <th>mutation_status</th>\n",
       "      <th>cnv_status</th>\n",
       "      <th>num_hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-3C-AAAU-01</th>\n",
       "      <td>class 4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-3C-AALI-01</th>\n",
       "      <td>class 4</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>one</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-3C-AALJ-01</th>\n",
       "      <td>class 4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-3C-AALK-01</th>\n",
       "      <td>class 4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-4H-AAAK-01</th>\n",
       "      <td>class 4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                class_name  mutation_status  cnv_status num_hits\n",
       "TCGA-3C-AAAU-01    class 4                0           0     none\n",
       "TCGA-3C-AALI-01    class 4                1           0      one\n",
       "TCGA-3C-AALJ-01    class 4                0           0     none\n",
       "TCGA-3C-AALK-01    class 4                0           0     none\n",
       "TCGA-4H-AAAK-01    class 4                0           0     none"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(cfg.distance_loss_info, 'rb') as f:\n",
    "    park_loss_info = pkl.load(f)\n",
    "    \n",
    "park_loss_info['TP53_BRCA'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate distance between means/medians for given gene + cancer type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(['both', 'none', 'one'],\n",
       " [('both', 'none'), ('both', 'one'), ('none', 'one')],\n",
       " [19, 491, 287],\n",
       " [5.290758839168914, 2.355077712145397, 3.8760241020922455])"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from scipy.spatial.distance import pdist, squareform\n",
    "\n",
    "def get_centroids_and_distance(identifier, info_df, centroid_method='mean'):\n",
    "    \n",
    "    groups = ['both', 'none', 'one']\n",
    "    group_combinations = list(it.combinations(groups, 2))\n",
    "    class_name = info_df['class_name'].values[0]\n",
    "    \n",
    "    # get expression data for samples\n",
    "    samples = info_df.index.intersection(data_df.index)\n",
    "    info_df = info_df.reindex(samples)\n",
    "    \n",
    "    # if one group has no samples, we have to make sure to assign it 0 count\n",
    "    class_counts = []\n",
    "    hit_class_counts = info_df.groupby('num_hits').count().class_name\n",
    "    for group in groups:\n",
    "        if group in hit_class_counts.index:\n",
    "            class_counts.append(hit_class_counts[group])\n",
    "        else:\n",
    "            class_counts.append(0)\n",
    "    \n",
    "    # group by number of hits, then calculate centroids\n",
    "    centroids_df = (data_df\n",
    "        .reindex(samples)\n",
    "        .merge(info_df['num_hits'], left_index=True, right_index=True)\n",
    "        .groupby('num_hits')\n",
    "    )\n",
    "    \n",
    "    if centroid_method == 'mean':\n",
    "        centroids_df = centroids_df.mean()\n",
    "    elif centroid_method == 'median':\n",
    "        centroids_df = centroids_df.median()\n",
    "    else:\n",
    "        raise NotImplementedError(\n",
    "            'centroid method {} not implemented'.format(centroid_method)\n",
    "        )\n",
    "    \n",
    "    # calculate distance between centroids\n",
    "    # make sure this is in the same order for each identifier, and\n",
    "    # handle NA distances here (if one group doesn't have any samples)\n",
    "    dists = pdist(centroids_df.values, metric='euclidean')\n",
    "    dist_combinations = list(it.combinations(hit_class_counts.index, 2))\n",
    "    ordered_dists = []\n",
    "    for cmb in group_combinations:\n",
    "        if cmb not in dist_combinations:\n",
    "            ordered_dists.append(np.nan)\n",
    "        else:\n",
    "            cmb_ix = dist_combinations.index(cmb)\n",
    "            ordered_dists.append(dists[cmb_ix])\n",
    "    \n",
    "    return (groups,\n",
    "            group_combinations,\n",
    "            class_counts,\n",
    "            ordered_dists)\n",
    "    \n",
    "get_centroids_and_distance('TP53_BRCA',\n",
    "                           park_loss_info['TP53_BRCA'],\n",
    "                           'median')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate centroid distance between \"hits\", per class\n",
    "\n",
    "Class 1 = look at both loss and gain (should be one-hit in neither)  \n",
    "Class 2 = only look at loss (should be one-hit here)  \n",
    "Class 3 = only look at gain (should be one-hit here)  \n",
    "Class 4 = look at both loss and gain (should be one-hit in both)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(433, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>both</th>\n",
       "      <th>none</th>\n",
       "      <th>one</th>\n",
       "      <th>class_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACVR1_UCEC</th>\n",
       "      <td>2</td>\n",
       "      <td>309</td>\n",
       "      <td>94</td>\n",
       "      <td>class 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR2A_COADREAD</th>\n",
       "      <td>3</td>\n",
       "      <td>264</td>\n",
       "      <td>96</td>\n",
       "      <td>class 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR2A_LIHC</th>\n",
       "      <td>0</td>\n",
       "      <td>134</td>\n",
       "      <td>29</td>\n",
       "      <td>class 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AJUBA_HNSC</th>\n",
       "      <td>5</td>\n",
       "      <td>132</td>\n",
       "      <td>65</td>\n",
       "      <td>class 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKT1_BRCA</th>\n",
       "      <td>5</td>\n",
       "      <td>650</td>\n",
       "      <td>142</td>\n",
       "      <td>class 1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 both  none  one class_name\n",
       "ACVR1_UCEC          2   309   94    class 1\n",
       "ACVR2A_COADREAD     3   264   96    class 1\n",
       "ACVR2A_LIHC         0   134   29    class 1\n",
       "AJUBA_HNSC          5   132   65    class 2\n",
       "AKT1_BRCA           5   650  142    class 1"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "class_counts_df = {}\n",
    "results_df = {}\n",
    "class_names = {}\n",
    "counts_columns = None\n",
    "results_columns = None\n",
    "\n",
    "# get distances for copy loss, for class 1/2/4 genes\n",
    "for identifier, loss_df in park_loss_info.items():\n",
    "    \n",
    "    class_name = loss_df.head(1).class_name.values[0]\n",
    "    if class_name == 'class 3':\n",
    "        continue\n",
    "        \n",
    "    results = get_centroids_and_distance(identifier, loss_df, 'mean')\n",
    "    \n",
    "    if counts_columns is None:\n",
    "        counts_columns = results[0]\n",
    "    else:\n",
    "        assert counts_columns == results[0]\n",
    "        \n",
    "    if results_columns is None:\n",
    "        results_columns = ['{}/{}'.format(i, j) for i, j in results[1]]\n",
    "            \n",
    "    class_names[identifier] = class_name\n",
    "    class_counts_df[identifier] = results[2]\n",
    "    results_df[identifier] = results[3]\n",
    "    \n",
    "class_counts_loss_df = (\n",
    "    pd.DataFrame(class_counts_df.values(),\n",
    "                 index=class_counts_df.keys(),\n",
    "                 columns=counts_columns)\n",
    "      .merge(pd.Series(class_names).rename('class_name'),\n",
    "             left_index=True, right_index=True)\n",
    ")\n",
    "    \n",
    "results_loss_df = (\n",
    "    pd.DataFrame(results_df.values(),\n",
    "                 index=results_df.keys(),\n",
    "                 columns=results_columns)\n",
    "      .merge(pd.Series(class_names).rename('class_name'),\n",
    "             left_index=True, right_index=True)\n",
    ")\n",
    "    \n",
    "print(class_counts_loss_df.shape)\n",
    "class_counts_loss_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(433, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>both/none</th>\n",
       "      <th>both/one</th>\n",
       "      <th>none/one</th>\n",
       "      <th>class_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACVR1_UCEC</th>\n",
       "      <td>4.716244</td>\n",
       "      <td>3.975686</td>\n",
       "      <td>1.581885</td>\n",
       "      <td>class 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR2A_COADREAD</th>\n",
       "      <td>5.484907</td>\n",
       "      <td>5.235206</td>\n",
       "      <td>0.809157</td>\n",
       "      <td>class 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR2A_LIHC</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.546988</td>\n",
       "      <td>class 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AJUBA_HNSC</th>\n",
       "      <td>2.757378</td>\n",
       "      <td>2.631804</td>\n",
       "      <td>0.824527</td>\n",
       "      <td>class 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKT1_BRCA</th>\n",
       "      <td>3.488913</td>\n",
       "      <td>3.118687</td>\n",
       "      <td>1.075147</td>\n",
       "      <td>class 1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 both/none  both/one  none/one class_name\n",
       "ACVR1_UCEC        4.716244  3.975686  1.581885    class 1\n",
       "ACVR2A_COADREAD   5.484907  5.235206  0.809157    class 1\n",
       "ACVR2A_LIHC            NaN       NaN  1.546988    class 1\n",
       "AJUBA_HNSC        2.757378  2.631804  0.824527    class 2\n",
       "AKT1_BRCA         3.488913  3.118687  1.075147    class 1"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(results_loss_df.shape)\n",
    "results_loss_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(384, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>both</th>\n",
       "      <th>none</th>\n",
       "      <th>one</th>\n",
       "      <th>class_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACVR1_UCEC</th>\n",
       "      <td>0</td>\n",
       "      <td>362</td>\n",
       "      <td>43</td>\n",
       "      <td>class 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR2A_COADREAD</th>\n",
       "      <td>2</td>\n",
       "      <td>320</td>\n",
       "      <td>41</td>\n",
       "      <td>class 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR2A_LIHC</th>\n",
       "      <td>2</td>\n",
       "      <td>133</td>\n",
       "      <td>28</td>\n",
       "      <td>class 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKT1_BRCA</th>\n",
       "      <td>0</td>\n",
       "      <td>557</td>\n",
       "      <td>240</td>\n",
       "      <td>class 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKT1_UCEC</th>\n",
       "      <td>0</td>\n",
       "      <td>347</td>\n",
       "      <td>58</td>\n",
       "      <td>class 1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 both  none  one class_name\n",
       "ACVR1_UCEC          0   362   43    class 1\n",
       "ACVR2A_COADREAD     2   320   41    class 1\n",
       "ACVR2A_LIHC         2   133   28    class 1\n",
       "AKT1_BRCA           0   557  240    class 1\n",
       "AKT1_UCEC           0   347   58    class 1"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "class_counts_df = {}\n",
    "results_df = {}\n",
    "class_names = {}\n",
    "counts_columns = None\n",
    "results_columns = None\n",
    "\n",
    "# get distances for copy gain, for class 1/3/4 genes\n",
    "for identifier, gain_df in park_gain_info.items():\n",
    "    \n",
    "    class_name = gain_df.head(1).class_name.values[0]\n",
    "    if class_name == 'class 2':\n",
    "        continue\n",
    "        \n",
    "    results = get_centroids_and_distance(identifier, gain_df, 'mean')\n",
    "    \n",
    "    if counts_columns is None:\n",
    "        counts_columns = results[0]\n",
    "    else:\n",
    "        assert counts_columns == results[0]\n",
    "        \n",
    "    if results_columns is None:\n",
    "        results_columns = ['{}/{}'.format(i, j) for i, j in results[1]]\n",
    "            \n",
    "    class_names[identifier] = class_name\n",
    "    class_counts_df[identifier] = results[2]\n",
    "    results_df[identifier] = results[3]\n",
    "    \n",
    "class_counts_gain_df = (\n",
    "    pd.DataFrame(class_counts_df.values(),\n",
    "                 index=class_counts_df.keys(),\n",
    "                 columns=counts_columns)\n",
    "      .merge(pd.Series(class_names).rename('class_name'),\n",
    "             left_index=True, right_index=True)\n",
    ")\n",
    "    \n",
    "results_gain_df = (\n",
    "    pd.DataFrame(results_df.values(),\n",
    "                 index=results_df.keys(),\n",
    "                 columns=results_columns)\n",
    "      .merge(pd.Series(class_names).rename('class_name'),\n",
    "             left_index=True, right_index=True)\n",
    ")\n",
    "    \n",
    "print(class_counts_gain_df.shape)\n",
    "class_counts_gain_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(384, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>both/none</th>\n",
       "      <th>both/one</th>\n",
       "      <th>none/one</th>\n",
       "      <th>class_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACVR1_UCEC</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.269823</td>\n",
       "      <td>class 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR2A_COADREAD</th>\n",
       "      <td>4.943627</td>\n",
       "      <td>4.712461</td>\n",
       "      <td>1.454358</td>\n",
       "      <td>class 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR2A_LIHC</th>\n",
       "      <td>2.812433</td>\n",
       "      <td>2.941939</td>\n",
       "      <td>1.276057</td>\n",
       "      <td>class 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKT1_BRCA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.499204</td>\n",
       "      <td>class 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKT1_UCEC</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.672928</td>\n",
       "      <td>class 1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 both/none  both/one  none/one class_name\n",
       "ACVR1_UCEC             NaN       NaN  1.269823    class 1\n",
       "ACVR2A_COADREAD   4.943627  4.712461  1.454358    class 1\n",
       "ACVR2A_LIHC       2.812433  2.941939  1.276057    class 1\n",
       "AKT1_BRCA              NaN       NaN  1.499204    class 1\n",
       "AKT1_UCEC              NaN       NaN  1.672928    class 1"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(results_gain_df.shape)\n",
    "results_gain_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot centroid distance results"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:mutation_fn]",
   "language": "python",
   "name": "conda-env-mutation_fn-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
